sqlg-core/src/main/java/org/umlg/sqlg/structure/SqlgDataSourceFactory.java

```diff
@@ -0,0 +1,20 @@
+package org.umlg.sqlg.structure;
+
+import org.apache.commons.configuration.Configuration;
+import javax.sql.DataSource;
+
+/**
+ * Created by petercipov on 27/02/2017.
+ */
+public interface SqlgDataSourceFactory {
+
+    SqlgDataSource setup(String driver, final Configuration configuration) throws Exception;
+    SqlgDataSource setupFromJndi(String jndiName, Configuration configuration) throws Exception;
```

Comment:
As for the suggestion - there is really no need for `SqlgGraph` to hold a reference to `SqlgDataSource`. All it really needs is a `DataSource` instance, because that's where it gets its connections from (notice that SqlgGraph only uses the sqlgdatasource to get the datasource by  `jdbc.url` - so this can be replaced by just directly getting the datasource in `SqlgGraph` constructor and then use that instead of `sqlgDatasource.get(this.getJdbcUrl())`).

So I'd keep the special handling for JNDI datasources - if `jdbc.url` starts with `jndi:`, obtain the datasource from JNDI. If it does not, I'd use your proposal for a datasource factory to be able to use different pooling mechanisms.